### ISSUE TITLE
Cannot get two controllers (intelligent and motion) to communicate with each other

### ISSUE THREAD

**Issue by console-beaver** [RoboMasterS1Challenge/robomaster_s1_can_hack#2]
We are trying to use your code (thanks alot for publishing it), but cannot get both controllers to negotiate. So intelligent controller keep sending CAN Bus packets to motion controller and motion controller's light stays red and never turned to blue.

Can you explain which CAN interface need to be connected to each controller? I.e. is it can1 connected to intelligent controller and can2 - to motion controller? Is it opposite? Thanks!

**Comment by console-beaver** [comment]
Tried to switch can buses around, i.e. can 1 goes to motion controller and can2 - to intelligent controller, but no worky.

Curious if anyone else besides the author got it working using this code?

**Comment by console-beaver** [comment]
Which S1 firmware you did test this code against?
Apparently, it doesn't work on our S1 which is running very old (original) firmware release...

**Comment by JohnieBraaf** [comment]
Yes, I can confirm it is working. I suspect that you're can bus timing is off. Make sure you clock it at 1Mbit.

**Comment by console-beaver** [comment]
Oh sorry, that was the very old question. I got it working too awhile ago.

**Comment by krvcz** [comment]
> Oh sorry, that was the very old question. I got it working too awhile ago.

Hi, how exactly did you solve the problem? Where should I set Can to 1 Mbit?

**Comment by console-beaver** [comment]
@krvcz my issue was related to CAN bus physical connection and not the baud rate, and could be different from your issue. Basically, I was trying to connect it to both intelligent controller and motion controller at the same time with both CAN1 and CAN2 interfaces. Now with the new code, the intelligent controller was eliminated completely, and you have to use CAN1 interface only (by default, unless you change the code to use CAN2 instead).

**Comment by krvcz** [comment]
> @krvcz my issue was related to CAN bus physical connection and not the baud rate, and could be different from your issue. Basically, I was trying to connect it to both intelligent controller and motion controller at the same time with both CAN1 and CAN2 interfaces. Now with the new code, the intelligent controller was eliminated completely, and you have to use CAN1 interface only (by default, unless you change the code to use CAN2 instead).

Thank you for reply. We stack with this project, so do you have any photos with connectivity ? Can we chat somewhere , for example gmail or Facebook etc.

Now i can’t download project, with error „no such or directory in lwip/otp.h „

**Comment by JohnieBraaf** [comment]
![20201111_222138](https://user-images.githubusercontent.com/1111346/98866015-8b799100-246c-11eb-8582-134fa5d872c9.jpg)

You only need 1 CAN bus is the latest version of the software. Still had it laying around so see pic.
In his previous version he indeed programmed it as a passthrough device. Not anymore.
It has been couple of month ago and I have not checked after that.

I also wrote a micropython version but it's not yet finished, it spins the wheels though. Let me know if interested.

**Comment by JohnieBraaf** [comment]
For wiring:
- Unplug the main unit, the one with the antennas
- connect Gimbal CAN bus (High, Low) wires to the CAN tranciever. No need to add ground
- power microcontroller and tranciever with 5V

Thats it.

PS. I also removed the termination resistor, not sure if needed, but I was also capturing the signal on the other microcontroller running micropython.

**Comment by krvcz** [comment]
> For wiring:
> 
> * Unplug the main unit, the one with the antennas
> * connect Gimbal CAN bus (High, Low) wires to the CAN tranciever. No need to add ground
> * power microcontroller and tranciever with 5V
> 
> Thats it.
> 
> PS. I also removed the termination resistor, not sure if needed, but I was also capturing the signal on the other microcontroller running micropython.


Thanks a lot ! Soon as possible i will inform you about progress or possible errors . Is there a version with 2 Can buses in the picture?

**Comment by JohnieBraaf** [comment]
Edit: yes, not sure why, but only 1 is actually communicating. Its the purple and grey wires. I think it is the 2 on the right.
Btw, to solve your error, you probably need to download or enable the middleware LWIP in STM-Cube IDE.

**Comment by krvcz** [comment]
> Edit: yes, not sure why, but only 1 is actually communicating. Its the purple and grey wires. I think it is the 2 on the right.
> Btw, to solve your error, you probably need to download or enable the middleware LWIP in STM-Cube IDE.

Ok, so before I have connected board with 2 CAN's, so now should i download older program with 2 CAN in code, right ?
if it will not working with 2 CAN i'm going to try with only one CAN

**Comment by JohnieBraaf** [comment]
You can have a look at my micropython implementation. It's probably much easier to understand. It's working but not finished.

[Micropython-S1.zip](https://github.com/RoboMasterS1Challenge/robomaster_s1_can_hack/files/5527505/Micropython-S1.zip)

**Comment by JohnieBraaf** [comment]
"Ok, so before I have connected board with 2 CAN's, so now should i download older program with 2 CAN in code, right ?
if it will not working with 2 CAN i'm going to try with only one CAN"

That version of the code has far less features. And it's actually great to not have to power the main unit. saves battery, so I would say better to have only 1 can. 

Also in Micropython I first had a passthrough desing but I was running into performance limitations.

**Comment by krvcz** [comment]
Unfortunately, at the moment I am limited to STM32. I will keep you posted on the progress.
Does your version of micropython use a Rasberry microcontroller?
I would like to have an alternative solution in case of stm problem.

**Comment by JohnieBraaf** [comment]
You can run micropython on STM32 boards, see: https://github.com/micropython/micropython/tree/master/ports/stm32/boards
You do need to compile a patched version, see my comments: https://github.com/micropython/micropython/issues/5508

It should also be easy to port to raspberry pi, but then you will need a USB can bus device, or a CAN BUS shield.

**Comment by krvcz** [comment]
One more question: What should be the boot order with ROS and does the boot order matter?
I am talking about version with STM

**Comment by JohnieBraaf** [comment]
I don't know, I was not using ROS. Also, there is no ROS in the firmware you are looking at.
The author is using his PC as a bridge and pushes UDP packages to his PC.
You do have this as an option: http://wiki.ros.org/rosserial

**Comment by krvcz** [comment]
hmmm... but in source files are ros files with packages .  So the Ros files that are in the source folders are not required?

**Comment by JohnieBraaf** [comment]
nope, he is running that on PC running ROS, its not used in the firmware for the STM32

**Comment by krvcz** [comment]
Sorry my bad I have't understood you before. Anyway, If I connect  everything (wires + connect with ROS server), then the LED should be blue?

**Comment by JohnieBraaf** [comment]
Yes, led blue on control board of Robomaster means it's working. Otherwise it will be red (no controller). Green means lab mode.

**Comment by Richard-best86** [comment]
Hi @JohnieBraaf , I saw your comment about micropython version. 
It interested me very much. You can tell me how to do it and whether I need to implement ROS?I need some tips.
Please reply.
best regards

**Comment by JohnieBraaf** [comment]
No I was not using ROS. Basically just use CAN functionality on micropython board. It's very simple. 
And I built a class to create CAN messages with CRC and such. It's basically a partial port of this repository.
I meant to use WIFI to send  commands, but back then the assyncio was limited, but I had it working at that point.
I believe in the meantime there were big improvements in de micropython project regarding assyncio to swith between CAN sending and WIFI receiving task.

If I find the time I will make a generic working example repository.

**Comment by krvcz** [comment]
Ok, so i was trying connect everything, but still i see red LED :(  For test i am using connection by ethernet . I don’t have any ideas to resolve problem . It’s version with one Can as you see and the newest source files.
![7B5F646A-6160-42C2-909A-D87370D8DF74](https://user-images.githubusercontent.com/74314536/99095912-feeae200-25d5-11eb-945d-907559cf4fe8.jpeg)
![66EF1E5C-66DF-414C-AF3A-78E52CC3CE9C](https://user-images.githubusercontent.com/74314536/99095939-0611f000-25d6-11eb-9342-8a3a26d27dae.jpeg)
![98B08EC7-F751-42D1-98E2-E58C29AB69B0](https://user-images.githubusercontent.com/74314536/99095964-0c07d100-25d6-11eb-93d8-80613fc98184.jpeg)
![3873F94A-662E-4A28-8FCC-79F9D4EA1F7D](https://user-images.githubusercontent.com/74314536/99095981-11651b80-25d6-11eb-84d7-293b61c0027e.jpeg)

**Comment by JohnieBraaf** [comment]
Here you see it is working like this. I modified the code when if was studying the firmware to send the wheels forward so I could debug. I don't know about ROS, sorry have not used it. 

Check your pins, I think you're wiring is different.

Do note that the moment that you power it on matters. It has something like a startup procedure. Number of initial messages.
So make sure the S1 is already running, then turn on microcontroller. 

![20201113_203458](https://user-images.githubusercontent.com/1111346/99114055-81809b00-25f0-11eb-85ba-83427419f6b0.jpg)

**Comment by krvcz** [comment]
Thank you a lot for your reply and help. So, I connect the same like you and everything is fine and now i can see blue LED. The only problem is with ROS . Ros server is working , ping stm work as well but robot doesn’t move. I don’ t know where is the problem. 
Could you say more about control by ps4 controller in micropython ? Is it wireless ? How connect ps4 controller? Etc. It’ s very important for me . I am very appreciate your help.

**Comment by JohnieBraaf** [comment]
@krvcz , @Richard-best86 , I have placed a working example here: https://github.com/JohnieBraaf/Robomaster-Micropython
Lets continue discussion there if you want. It currently only turns on all leds and provides for blue led on the control board.
I will add some basic commands later. B.t.w., micropython seems to support the ethernet cable as well.
